PC/winreg.c

```diff
@@ -579,6 +579,9 @@ Py2Reg(PyObject *value, DWORD typ, BYTE **retDataBuf, DWORD *retDataSize)
             }
             else {
                 DWORD d = PyLong_AsUnsignedLong(value);
+                if ((d == (DWORD)(-1)) && PyErr_Occurred()) {
+                    return FALSE;
```

Comment:
I don't know this part of the code base, _but_ it does look like this leaks the memory allocated on line 570. Ditto for the change at line 603.

Reply:
It doesn't leak memory. The problem is that if an error occurred in PyLong_AsUnsignedLong (like passing a negative python integer), PyLong_AsUnsignedLong returns -1 and the code here think -1 is the value to be set and sets it. This PR solves that problem.

Reply:
Looks like a leak to me; the caller does not free the memory if `FALSE` is returned :)

Reply:
cc. @eryksun

Reply:
Agreed - `retDataBuf` isn't getting freed in that case.

Reply:
Ooops. Didn't notice that. Just freeing the memory using `free` will do, right?

Reply:
I'm fixing that. Wait.

Reply:
@pfmoore There seems to be inconsistent behaviour for the various cases; for `REG_EXPAND_SZ` memory is only allocated if `value == Py_None`. Don't know if it matters, though :)

Reply:
@shreyanavigyan Either free the memory explicitly before returning, or just move the conversion up before the allocation. For my own code, I prefer deferring allocation as late as possible in order to make it easier to clean up.

Reply:
Also, this part of the code base seems to use the deprecated `PyMem_FREE` iso. `PyMem_Free`.

Reply:
> Also, this part of the code base seems to use the deprecated `PyMem_FREE`/`PyMem_NEW` iso. `PyMem_Free`/`PyMem_New`.

I'll open a different issue and PR for that if needed.

Reply:
See GH-23586 for a previous cleanup, @shreyanavigyan.

Reply:
> move the conversion up before the allocation. 

Seems like a good idea to me. For example:

```C
        case REG_DWORD: {
            DWORD d = 0;
            if (PyLong_Check(value)) {
                d = PyLong_AsUnsignedLong(value);
                if (d == (DWORD)(-1) && PyErr_Occurred()) {
                    return FALSE;
                }
            } else if (value != Py_None) {
                return FALSE;
            }
            *retDataBuf = (BYTE *)PyMem_NEW(DWORD, 1);
            if (*retDataBuf == NULL) {
                PyErr_NoMemory();
                return FALSE;
            }
            memcpy(*retDataBuf, &d, sizeof(DWORD));
            *retDataSize = sizeof(DWORD);
            break;
        }
        case REG_QWORD: {
            DWORD64 d = 0;
            if (PyLong_Check(value)) {
                d = PyLong_AsUnsignedLongLong(value);
                if (d == (DWORD64)(-1) && PyErr_Occurred()) {
                    return FALSE;
                }
            } else if (value != Py_None) {
                return FALSE;
            }
            *retDataBuf = (BYTE *)PyMem_NEW(DWORD64, 1);
            if (*retDataBuf == NULL) {
                PyErr_NoMemory();
                return FALSE;
            }
            memcpy(*retDataBuf, &d, sizeof(DWORD64));
            *retDataSize = sizeof(DWORD64);
            break;
        }
```

Reply:
Should I implement this? (I've already added a commit to free up retDataBuf before returning FALSE. Is it worth changing the whole code control flow?)

Reply:
IMO it's a better solution (avoids unneeded malloc/free). I leave it for Eryk and Paul to decide :)

Reply:
I agree. There's no requirement to touch as few lines of code as possible with a change, so let's use the cleanest approach we can to get the required behaviour.

Reply:
Should I proceed to change the code?

Reply:
Looking at the existing code again. It seems changing it would be a better approach. I'm applying the changes that @eryksun suggested.